### ISSUE TITLE
Update to borg 1.2.6 and add automatic migration steps

### ISSUE THREAD

**Issue by MartinX3** [borgbase/vorta#1794]
See https://github.com/borgbackup/borg/blob/1.2.6/docs/changes.rst#pre-125-archives-spoofing-vulnerability-cve-2023-36811

Or add a warning on first Vorta startup preventing a `borg check` without following them.

**Comment by m3nu** [comment]
Our release cycle isn't fast enough to address this ad-hoc. And many users will not run 1.2.6 yet. So Vorta may not be the right place to address this. We only aim to cover the most frequent commands and not 100%.

**Comment by MartinX3** [comment]
But doesn't Vorta run periodic checks on the repo by default?
This could kill the backups if I read the warning correctly.

**Comment by real-yfprojects** [comment]
Vorta doesnbt upgrade borg. Consequently the user is responsible with following the proper upgrade steps. However we do provide a flatpak and homebrew package with bundled borg. Upgrading borg in those needs special attention.

**Comment by MartinX3** [comment]
Oh, I am sorry.
I just saw 1.2.4 as the borg version in vorta so I thought it was bundled.
But all I need to do was to restart vorta.

**Comment by real-yfprojects** [comment]
> Oh, I am sorry. I just saw 1.2.4 as the borg version in vorta so I thought it was bundled. But all I need to do was to restart vorta.

Yeah, the version is only checked on restart. Checking the version regularly while running wouldn't make much sense, would it.

**Comment by real-yfprojects** [comment]
How should Vorta know that you updated your borg version. Running the check every time Vorta starts, doesn't seem viable to me.

**Comment by Hofer-Julian** [comment]
For the flatpak, I keep borg pinned to 1.24 at the moment: https://github.com/flathub/com.borgbase.Vorta/blob/a08e83a9da5c34c5040d59b9b316c16c007f72b8/dependencies/borgbackup.json#L10

I assume this is still the way to go until this (or a similar) feature is implemented?

**Comment by sten0** [comment]
This will affect a couple thousand users as soon as this April, when Ubuntu 22.04 is released with Borg 1.2.7, and it already affects Ubuntu 23.10 users.  Yes, Debian testing or unstable users are also currently affected.  It is not possible to for affected users to downgrade Borg.

If "borg check" will cause data loss for some cases, then these cases should be handled so as not to cause data loss.  Trust in backup software is like trust in a file system: people don't give second chances.  It is better for backup software to say "sorry, you're using an unsupported version of borg" than to cause data loss.  GUI users may need a borg repo upgrade wizard function to navigate this sort of issue.

**Comment by ThomasWaldmann** [comment]
- no
- guess one could use the steps listed in the cve / upgrade notes of borg to detect if there is are upgrade steps needed. For borg > 1.2.4 manifest and archives all must have a TAM (including .checkpoint archives for completeness).
- there could be also a convenience single-click "check and upgrade" action that trusts everything currently in the repo. that's not what paranoid / very cautious users want, but guess it is good enough for most users.

**Comment by m3nu** [comment]
I could add a "TAM Upgrade" button and command fairly quickly together with a link for more details. Or we only watch the logs for the relevant error and display an alert with a link for more details.

OTOH, the macOS release has included a newer Borg version since last year and I don't recall getting many questions. Also not to our support. We pointed the issue out twice in the monthly BorgBase newsletter. So it's possible this doesn't affect as many users? I believe I only had to upgrade one of my repos, which was missing the TAM for one archive only. So this case may be the most common?

> If there are no tam:none archives left at this point, you can skip this step.

**Comment by ramcq** [comment]
I found this ticket as I use the Vorta Flatpak - I am intermittently encountering https://github.com/borgbackup/borg/issues/6687 so I was looking if I could update the Flatpak to 1.2.7 which led me here. I think Vorta could potentially aid the migration by tracking, per archive, a boolean indicating whether the migration has been carried out - for existing archives where this flag is false, it could automatically check TAM validity and flip it to true, which seems to be a common case. If the TAM is not valid it can notify the user with a warning / link to the advisory and an explicit step to do the upgrade with a warning. Once an archive has been migrated, or if it was newly created, it should simply enforce that the TAM is valid and not offer the migration path - this would avoid Vorta ever allowing the migration to be carried out more than once and any TAM invalidity can then be reported with the appropriate severity to the user.

**Comment by hozza** [comment]
I've run into this issue after doing an initial backup (to borgbase) with flatpak Vorta (using bundled borg 1.2.4). 

I get errors when attempting to setup automated regular backups using [borgmatic-collective/docker-borgmatic](https://github.com/borgmatic-collective/docker-borgmatic) docker container.

It turns out the docker-borgmatic container uses borg 1.2.7, thankfully I was using an append-only ssh key so helpfully no data loss from the check run (although I'm not sure on that?). I'll disable auto backups with this until Vorta implements a "fix/upgrade repo" option or updates the bundled borg I guess?

(I use Vorta manually with a privileged ssh key on a trusted device for repo prune/management, and an append only ssh key on the actual backup client device in case of it being compromised and attempting to destroy it's own backups for ransom or whatever.) 

Is this something that Borgbase should alert users about on the dashboard/by email?

_This self corrupting version issue feels a little like Captin Janeway realising a nanovirus into the [central plexus](https://memory-alpha.fandom.com/wiki/Central_plexus) of a [class-4 tactical cube](https://memory-alpha.fandom.com/wiki/Borg_tactical_cube). 🦾_

**Comment by ThomasWaldmann** [comment]
@hozza The steps you should do for your repo are described at the top of the borg 1.2.7 changelog. You could just copy&paste and slightly modify these for your repo URL.

Guess nothing bad happens, if you don't run `borg check --repair` until then with borg > 1.2.4.

**Comment by ThomasWaldmann** [comment]
I had some time and added some helper commands to simplify the upgrade steps.

I would appreciate some review and practical testing of https://github.com/borgbackup/borg/pull/8159 .

The new commands are for checking whether something needs to be done and exit with rc=0 if all looks OK and rc=1 if something TAM-related needs fixing.

Of course these new commands will only be available with borg >= 1.2.8.

**Comment by sten0** [comment]
Belated thank you @ThomasWaldmann.  How is the testing going?

**Comment by ThomasWaldmann** [comment]
.

**Comment by m3nu** [comment]
We can keep this open. But I've seen very few people with this problem and I don't feel too good about scripting this and ignoring potential edge cases. And it's one-off so the code will be redundant at some point.

**Comment by sten0** [comment]
@TW I've now had a chance to test the new easy-to-use CVE-2023-36811
mitigation process from Borg 1.2.8 on a bunch of clients, and it worked
perfectly everywhere.

@Manu The newest backup that I've seen with this problem is from 
2016-12-13.  The only Ubuntu (or derivatives) user I know isn't yet a 
Borg/Vorta user, but there's a reasonable expectation that Ubuntu (and 
derivative) LTS users were using an ancient (<1.0.9) Borg version as late 
as 2018 or 2020.  The only reason I use Debian & Ubuntu as a point of 
reference is because it's known which package versions users use long-term, 
so it's possible to have some idea of risk/impact.  Then a Borg user would 
have needed to become a Vorta user.

Both of us can claim that we're not responsible for users who imported
their old repositories, and Vorta has only publically existed since 2018
(as far as I know), and only a pre-2017 release of Borg can trigger this
issue.

That said, this won't be the last time repositories need to be
revalidated.  It might be ten or fifteen years away, or it might only be
a couple of years, but it is pretty much certain that borg will
deprecate unsafe crypto and hash algorithms over time.  Thus particular
cases will eventually become useless (note: one off per user for
unbounded number of users), but the entry point will remain useful.

```
Before running a "borg check (and borg check --repair)"
  Loop over issues
    When Vorta does not have proof of having checked for issue "x"?
      when x="CVE-2023-36811"
        borg upgrade --show-rc --check-tam
          if (rc != 0)
            refer user to Borg documentation
            abort now
          do_more?
      when x="next-issue"
…etc
```

That said, it's tricky when checking the repo state requires multiple
calls to borg that change the state of the repository.  That's really
what this issue is about I think:  How to mitigate destructive repo
checks.

**Comment by ThomasWaldmann** [comment]
IIRC also newer borg versions before 1.2.5 can trigger that issue, e.g. if you use rename or recreate (there was a bug in these function that did not correctly re-authenticate / re-compute the TAM for the affected archive).
